src/herder/EnvelopeItemMap.cpp

```diff
@@ -0,0 +1,165 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "herder/EnvelopeItemMap.h"
+
+namespace stellar
+{
+
+using xdr::operator<;
+
+void
+EnvelopeItemMap::add(SCPEnvelope const& envelope,
+                     std::vector<ItemKey> const& items)
+{
+    for (auto const& itemKey : items)
+    {
+        add(envelope, itemKey);
+    }
+}
+
+void
+EnvelopeItemMap::add(SCPEnvelope const& envelope, ItemKey itemKey)
+{
+    mItemEnvelopes[itemKey].insert(envelope);
+    mEnvelopeItems[envelope].insert(itemKey);
+}
+
+std::set<ItemKey>
+EnvelopeItemMap::remove(SCPEnvelope const& envelope)
+{
+    auto result = std::set<ItemKey>{};
+    auto it = mEnvelopeItems.find(envelope);
+    if (it == std::end(mEnvelopeItems))
+    {
+        return result;
+    }
+
+    for (auto const& itemKey : it->second)
+    {
+        if (removeEnvelopeLink(envelope, itemKey))
+        {
+            // it was last envelope waiting for this item
+            result.insert(itemKey);
+        }
+    }
+
+    mEnvelopeItems.erase(it);
+    return result;
+}
+
+std::set<SCPEnvelope>
+EnvelopeItemMap::remove(ItemKey itemKey)
+{
+    auto result = std::set<SCPEnvelope>{};
+    auto it = mItemEnvelopes.find(itemKey);
+    if (it == std::end(mItemEnvelopes))
+    {
+        return result;
+    }
+
+    for (auto const& envelope : it->second)
+    {
+        if (removeItemLink(envelope, itemKey))
+        {
+            // it was last item this envelope was waiting for, we can add it to
+            // result
+            result.insert(envelope);
+        }
+    }
+
+    mItemEnvelopes.erase(it);
+    return result;
+}
+
+std::set<ItemKey>
+EnvelopeItemMap::removeIf(std::function<bool(SCPEnvelope const&)> cond)
+{
+    auto result = std::set<ItemKey>{};
+    for (auto it = std::begin(mEnvelopeItems); it != std::end(mEnvelopeItems);)
+    {
+        if (cond(it->first))
+        {
+            it = mEnvelopeItems.erase(it);
```

Comment:
I already have iterator for this envelope, and using `EnvelopeItemMap::remove(SCPEnvelope const& envelope) ` would mean looking for it again, which would be inefficient.

It would also mean that each item for each of envelopes that passes `cond` is looked for multiple times, so a lot of unneccessary std::map::find calls.